### PR TITLE
Custom Structs for ACSets

### PR DESCRIPTION
**Pull request by olynch**
This is a draft for generating custom structs for ACSets, related to #375. Currently, if pres is the presentation for `WeightedGraph`, then `struct_acset(:WeightedGraph, pres, [:src,:tgt])` produces

``` julia
quote
    #= /g/Catlab.jl/src/categorical_algebra/StructAcsets.jl:42 =#
    struct WeightedGraph{Weight}
        #= /g/Catlab.jl/src/categorical_algebra/StructAcsets.jl:43 =#
        tables::NamedTuple{(:V, :E), Tuple{Table{NamedTuple{(), Tuple{}}}, Table{NamedTuple{(:src, :tgt, :weight), Tuple{Int64, Int64, Weight}}}}}
    end
    #= /g/Catlab.jl/src/categorical_algebra/StructAcsets.jl:46 =#
    schema(::Type{<:WeightedGraph}) = begin
            #= /g/Catlab.jl/src/categorical_algebra/StructAcsets.jl:46 =#
            Presentation{Schema,Symbol}(Catlab.Theories.FreeSchema, (Ob = Catlab.Theories.FreeSchema.Ob{:generator}[V, E], Hom = Catlab.Theories.FreeSchema.Hom{:generator}[src, tgt], Data = Catlab.Theories.FreeSchema.Data{:generator}[Weight], Attr = Catlab.Theories.FreeSchema.Attr{:generator}[weight]), Dict(:src => (:Hom => 1),:weight => (:Attr => 1),:V => (:Ob => 1),:E => (:Ob => 2),:tgt => (:Hom => 2),:Weight => (:Data => 1)), Pair[])
        end
end
```

I still have to add indices and make this into a macro, but as a proof of concept this is totally doable. I'm now having vague recollections of discussing this before; did we decide that something made this undesirable? I feel like it really might solve a lot of our problems.

### CODE REVIEW

**Comment by olynch** [comment]
The good news is that I think most of the accessing code doesn't even need to change that much! We even get to clean it up, because we can work with the presentation object directly instead of the clumsy `CatDesc` and `AttrDesc`.

**Comment by jpfairbanks** [comment]
Yes, and functions that are constant given type of the input argument like your schema function, are subject to constant propagation at compile time. So no runtime performance penalty is introduced there.

**Comment by olynch** [comment]
Well, schema is only called a `@generation` time, so it doesn't really matter either way :D

**Comment by jpfairbanks** [comment]
I can see an application where you would want to avoid generated functions and use a more generic implementation to avoid compile costs. Like if you used ACSets to interface with a DB and every query result came back as an ACSet with a different schema. You would want a generic implementation for all schemas rather than specializing on each schema so that the compiler wasn’t invoked for every query. But that application is far away and this refactor will help with that by making subtyping less janky.

**Comment by olynch** [comment]
> I can see an application where you would want to avoid generated functions and use a more generic implementation to avoid compile costs. Like if you used ACSets to interface with a DB and every query result came back as an ACSet with a different schema. You would want a generic implementation for all schemas rather than specializing on each schema so that the compiler wasn’t invoked for every query. But that application is far away and this refactor will help with that by making subtyping less janky.

Yes, this makes a lot of sense, and this is exactly what I mean when I say "custom implementations". I.e.; we can have different implementations of the basic operations, including generic ones.

**Comment by olynch** [comment]
OK, while we're at it, we might as well clean up the interface for declaring an acset. Really, it should be as simple as:

```julia
@acset_type WeightedGraph{T} begin
  V::Ob
  E::Ob with {src::V,weight::T}
  tgt::Hom(E,V)
end
```

I.e., we support both declaring homs in-line and separately, and data objects are automatically declared through the generic type. This also obviates the annoying requirement to have both "WeightedGraph" and "TheoryWeightedGraph". We will just have "WeightedGraph" and "schema(WeightedGraph)".

File: src/categorical_algebra/StructAcsets.jl [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+  name::Symbol
+  dom::Symbol
+  codom::Symbol
+  idxed::Bool
+end
+
+""" Same information as Presentation{Schema}, but without equations,
+and in a much more concrete/easily accessible format.
+"""
+struct ParsedSchema
```
Comment (jpfairbanks) [inline src/categorical_algebra/StructAcsets.jl:68]: I guess we can't use an ACSet for this... too bad the CT primordial soup can't be implemented as a soup.

File: src/theories/ProductSchema.jl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  """
+  compose(f::Hom(A,B), g::Attr(B,X))::Attr(A,X) ⊣ (A::Ob, B::Ob, X::Data)
+
+  (compose(f, compose(g, a)) == compose(compose(f, g), a)
+    ⊣ (A::Ob, B::Ob, C::Ob, X::Data, f::Hom(A,B), g::Hom(B,C), a::Attr(C, X)))
+  compose(id(A), a) == a ⊣ (A::Ob, X::Ob, a::Attr(A,X))
+end
+
+@syntax FreeSchema{ObExpr,HomExpr,DataExpr,AttrExpr} ProductSchema begin
+  # should have a normal representation for precompose of a morphism + a generator attribute
```
Comment (jpfairbanks) [inline src/theories/ProductSchema.jl:17]: this normal form should just be a list of homs and the optional attribute name right? So the syntax should associate things into a list and then when you compose two expressions, you make sure that only the last one has a non-null attribute?

Comment (jpfairbanks) [reply src/theories/ProductSchema.jl:17]: @epatters, can Catlab support custom structs for the syntax system? That seems like the way to go for making Operads fit in to the whole approach.

**Review by jpfairbanks — COMMENTED** [review]
I haven't fully processed the parsing code, but I don't see any red flags, @epatters what do you think?

**Comment by olynch** [comment]
@epatters, we need to figure out naming for `CSets.jl`/`Acsets.jl`. Maybe what we currently call `Acsets.jl` should be `AcsetInterface.jl`?

**Comment by olynch** [comment]
One big todo here is to reimplement the tables.jl functionality, now that this doesn't use a tables.jl thing internally.

**Comment by olynch** [comment]
Also need to do `unique_index`

**Comment by olynch** [comment]
And the `@acset` macro, and `map`. Soo much acset infrastructure, this better all be worth it...

**Review by epatters — APPROVED** [review]

**Pull request merged**